lmdeploy/pytorch/models/deepseek_v2.py

```diff
@@ -0,0 +1,373 @@
+            # (num_heads, nope_head_dim//2, 2, input_dim)
+            new_w_pe = w_pe.unflatten(1, (-1, 2))
+            # (num_heads, nope_head_dim, input_dim)
+            new_w_pe = new_w_pe.transpose(1, 2).flatten(1, 2)
+            weight[:, pe_dim_offset:] = new_w_pe
+
+        # prevent shuffle before apply rotary embedding
+        if self.q_lora_rank is None:
+            __update_pe(self.q_proj, self.q_head_dim, qk_nope_head_dim)
+        else:
+            __update_pe(self.q_b_proj, self.q_head_dim, qk_nope_head_dim)
+        kv_dim = self.kv_lora_rank + self.qk_rope_head_dim
+        __update_pe(self.kv_a_proj_with_mqa, kv_dim, self.kv_lora_rank)
+
+        kv_b_proj = self.kv_b_proj
+        w_kc, w_vc = kv_b_proj.weight.unflatten(
+            0, (-1, qk_nope_head_dim + v_head_dim)).split(
+                [qk_nope_head_dim, v_head_dim], dim=1)
+
+        self.register_parameter('w_kc',
+                                torch.nn.Parameter(w_kc, requires_grad=False))
+        w_vc = w_vc.transpose(1, 2).contiguous()
+        self.register_parameter('w_vc',
+                                torch.nn.Parameter(w_vc, requires_grad=False))
+
+        delattr(self, 'kv_b_proj')
+        gc.collect()
+
+    @classmethod
+    def _distribute_output_fn(cls, outputs, **kwargs):
+        """Distribution output hook."""
+        dist.all_reduce(outputs[0])
+        return outputs
+
+    def _contiguous_batching_forward_impl(
+        self,
+        hidden_states: torch.Tensor,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Any] = None,
+        world_size: int = 1,
+    ):
+        """forward impl."""
+        context = self.context.context
+        kv_seq_length = context.kv_seq_length
+        q_seq_length = context.q_seq_length
+        q_start_loc = context.q_start_loc
+        block_offsets = context.block_offsets
+        max_q_seq_length = context.max_q_seq_length
+        max_kv_seq_length = context.max_kv_seq_length
+        num_heads = self.num_heads // world_size
+        q_len = hidden_states.size(1)
+
+        def __q_proj(hidden_states, nope_size: int, pe_size: int):
+            """q proj."""
+            query_states = hidden_states.new_empty(q_len, num_heads,
+                                                   nope_size + pe_size)
+
+            if self.q_lora_rank is None:
+                q = self.q_proj(hidden_states)
+            else:
+                q = self.q_b_proj(
+                    self.q_a_layernorm(self.q_a_proj(hidden_states)))
+            q = q.view(q_len, num_heads, self.q_head_dim)
+            # q_pe: (q_len, num_heads, qk_rope_head_dim)
+            q_nope, q_pe = torch.split(
+                q, [self.qk_nope_head_dim, self.qk_rope_head_dim], dim=-1)
+            # q_nope: (q_len, num_heads, kv_lora_rank)
+            q_nope_out = query_states[..., :nope_size]
+            torch.bmm(q_nope.transpose(0, 1),
+                      self.w_kc,
```

Comment:
Hi @grimoire, have you considered how to absorb the `w_kc` into q_proj's weights offline instead of computing it in the online forward step?

Reply:
Merge weight of bmm required too many memory.

Reply:
Like this?

```python
import torch

def naive(a, w0, w1):
    num_heads = w1.size(0)
    mem_size = w0.numel() * w0.element_size() + w1.numel() * w0.element_size()
    print(f'naive memory size: {mem_size}')
    out = a@w0
    out = out.unflatten(1, (num_heads, -1)) # (seq, num_heads, w0_size)
    out = torch.bmm(out.transpose(0, 1), w1)
    return out.transpose(0, 1)


def fused(a, w0, w1):
    num_heads = w1.size(0)
    w0 = w0.unflatten(1, (num_heads, -1))
    w2 = torch.bmm(w0.transpose(0, 1), w1)
    w2 = w2.transpose(0, 1).flatten(1, 2)
    mem_size = w2.numel() * w2.element_size()
    print(f'fused memory size: {mem_size}')
    out = a@w2
    out = out.unflatten(-1, (num_heads, -1))
    return out


def main():
    in_size = 8
    num_heads = 16
    w0_size = 128
    w1_size = 512
    a = torch.rand(3, in_size)
    w0 = torch.rand(in_size, num_heads * w0_size)
    w1 = torch.rand(num_heads, w0_size, w1_size)

    naive_out = naive(a, w0, w1)
    fused_out = fused(a, w0, w1)

    torch.testing.assert_close(naive_out, fused_out)


if __name__ == '__main__':
    main()

```